python/taichi/testing.py

```diff
@@ -0,0 +1,108 @@
+import taichi as ti
+
+print('[Taichi] loading test module')
+
+
+## Helper functions
+def approx(expected, **kwargs):
+    '''Tweaked pytest.approx for OpenGL low percisions'''
+    class boolean_integer:
+        def __init__(self, value):
+            self.value = value
+
+        def __eq__(self, other):
+            return bool(self.value) == bool(other)
+
+        def __ne__(self, other):
+            return bool(self.value) != bool(other)
+
+    if isinstance(expected, bool):
+        return boolean_integer(expected)
+
+    if ti.cfg.arch == ti.opengl:
+        kwargs['rel'] = max(kwargs.get('rel', 1e-6), 1e-3)
+
+    return pytest.approx(expected, **kwargs)
+
+
+def allclose(x, y, **kwargs):
+    '''Same as: x == approx(y, **kwargs)'''
+    return x == approx(y, **kwargs)
+
+
+def make_temp_file(*args, **kwargs):
+    '''Create a temporary file'''
+    import os
+    from tempfile import mkstemp
+    fd, name = mkstemp(*args, **kwargs)
+    os.close(fd)
+    return name
+
+
+## Pytest options
+def get_conftest(_glbs):
+    import pytest
+
+    @pytest.fixture(params=ti.supported_archs(), ids=ti.core.arch_name)
+    def taichi_archs(request):
+        marker = request.node.get_closest_marker('taichi')
+        req_arch = request.param
+
+        def ti_init(arch=None, exclude=None, require=None, **options):
+            if arch is None:
+                arch = []
+            if exclude is None:
+                exclude = []
+            if require is None:
+                require = []
+            if not isinstance(arch, (list, tuple)):
+                arch = [arch]
+            if not isinstance(exclude, (list, tuple)):
+                exclude = [exclude]
+            if not isinstance(require, (list, tuple)):
+                require = [require]
+            if len(arch) == 0:
+                arch = ti.supported_archs()
+
+            if (req_arch not in arch) or (req_arch in exclude):
+                raise pytest.skip(f'Arch={req_arch} not included in test')
+
+            if not all(
+                    ti.core.is_extension_supported(req_arch, e)
+                    for e in require):
+                raise pytest.skip(
+                    f'Arch={req_arch} some extension not satisfied')
+
+            ti.init(arch=req_arch, **options)
+
+        ti_init(*marker.args, **marker.kwargs)
+        yield
+
+    _lcls = dict(locals())
+    for _k, _v in _lcls.items():
+        if not _k.startswith('_'):
+            _glbs[_k] = _v
```

Comment:
Manipulating `locals()` and `globals()` is confusing and we should avoid as much as we can.

IIRC the code here is to add `taichi_archs` to the global scope. Is this necessary? It seems that we only need to register the method to `pytest` and later refer to the method using the `taichi_archs` string.

Reply:
No, the stupid pytest will only search `conftest.py` for functions decorated by `@pytest.fixture`.
It's **not the `pytest.fixture` registered `taichi_archs`, but the pytest runner read the `globals()` of `conftest.py`**. Removing this will cause:
```
______________ ERROR at setup of test_allclose_rel_reordered2[<lambda>-0.1] ________________
[gw1] linux -- Python 3.8.3 /usr/bin/python3
file /home/bate/Develop/taichi/tests/python/test_test.py, line 107
  @pytest.mark.parametrize('x', [0.1, 3])
  @pytest.mark.parametrize('allclose',
                           [ti.allclose, lambda x, y: x == ti.approx(y)])
  @ti.test()
  def test_allclose_rel_reordered2(x, allclose):
E       fixture 'taichi_archs' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, testrun_uid, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, worker_id
>       use 'pytest --fixtures [testpath]' for help on them.

/home/bate/Develop/taichi/tests/python/test_test.py:107
================================== short test summary info ==================================
ERROR tests/python/test_test.py::test_arch_list_cpu
ERROR tests/python/test_test.py::test_arch_cpu
ERROR tests/python/test_test.py::test_all_archs
ERROR tests/python/test_test.py::test_exclude_list_cpu
ERROR tests/python/test_test.py::test_arch_opengl
ERROR tests/python/test_test.py::test_exclude_cpu
ERROR tests/python/test_test.py::test_multiple_archs
ERROR tests/python/test_test.py::test_allclose_rel[allclose-0.1]
ERROR tests/python/test_test.py::test_require_extensions_2
ERROR tests/python/test_test.py::test_init_args
ERROR tests/python/test_test.py::test_require_extensions_1
ERROR tests/python/test_test.py::test_allclose_rel[<lambda>-0.1]
ERROR tests/python/test_test.py::test_allclose_rel[allclose-3]
ERROR tests/python/test_test.py::test_allclose_rel[<lambda>-3]
ERROR tests/python/test_test.py::test_allclose_rel_reordered1[allclose-0.1]
ERROR tests/python/test_test.py::test_allclose_rel_reordered1[<lambda>-0.1]
ERROR tests/python/test_test.py::test_allclose_rel_reordered1[allclose-3]
ERROR tests/python/test_test.py::test_allclose_rel_reordered1[<lambda>-3]
ERROR tests/python/test_test.py::test_allclose_rel_reordered2[allclose-3]
ERROR tests/python/test_test.py::test_allclose_rel_reordered2[<lambda>-3]
ERROR tests/python/test_test.py::test_allclose_rel_reordered2[allclose-0.1]
ERROR tests/python/test_test.py::test_allclose_rel_reordered2[<lambda>-0.1]
==================================== 22 errors in 1.33s =====================================
```

Reply:
OK. Please be courteous and avoid using aggressive words such as "stupid". There is no need to be rude here.